File: rxdownloadlibrary/src/main/java/rx/downloadlibrary/DownloadProgressInterceptor.java
package rx.downloadlibrary;


import java.io.IOException;

import okhttp3.Interceptor;
import okhttp3.Response;
import rx.RxBus;

public class DownloadProgressInterceptor implements Interceptor, DownloadProgressListener {

    public DownloadProgressInterceptor() {}

    @Override
    public Response intercept(Chain chain) throws IOException {
        Response originalResponse = chain.proceed(chain.request());
        return originalResponse.newBuilder()
                .body(new DownloadProgressResponseBody(originalResponse.request().url().toString(), originalResponse.body(), this))
                .build();
    }

    @Override
    public void update(String url, long bytesRead, long contentLength, boolean done) {
        RxBus.publish(DownloadEvent.newEvent(url, contentLength, bytesRead));
    }
}

File: rxdownloadlibrary/src/main/java/rx/downloadlibrary/DownloadProgressInterceptor.kt
package rx.downloadlibrary


import okhttp3.Interceptor
import okhttp3.Response
import rx.RxBus

class DownloadProgressInterceptor : Interceptor, DownloadProgressListener {

    override fun intercept(chain: Interceptor.Chain): Response {
        val originalResponse = chain.proceed(chain.request())
        val url = chain.request().url().toString()
        return originalResponse.newBuilder()
                .body(DownloadProgressResponseBody(url, originalResponse.body(), this))
                .build()
    }

    override fun update(url: String, bytesRead: Long, contentLength: Long) {
        val downloadable = Downloadable(url)
        downloadable.contentLength = contentLength
        downloadable.byteRead = bytesRead
        RxBus.publish(downloadable)
    }
}